### DISCUSSION TITLE
Fused Jasper block

### DISCUSSION THREAD

**Discussion by fedorovgv** [NVIDIA-NeMo/NeMo#4973] [General]
Hi everyone! In [Jasper paper](https://arxiv.org/pdf/1904.03288.pdf) i found this lines: 

> Each sub-block can be fused into a single GPU kernel: dropout is not used at inference-time and is
> eliminated, batch norm can be fused with the preceding convolution, ReLU clamps the result, and 
> residual summation can be treated as a modified bias term in this fused operation.

But in the jasper block code in nemo, I didn't see this sub-block merging completely into the core. Do I understand correctly that the line "Each sub-block can be fused" means that the operations in the sub-block are fused separately (conv->batch_norm->relu->dropout and conv fused, then the results are returned in batch_norm, which are also fused, etc.), not the entire sub-block?

No attempt is made to completely fused the subblock.

https://github.com/NVIDIA/NeMo/blob/5d5c0fbff7ebe5c5991fb99a1cfeb57d5014ad42/nemo/collections/asr/parts/submodules/jasper.py#L974-L1051

Can someone help me. I'm thinking, maybe these are some other details responsible for the fusion of the entire sub-block? or does "fused" mean that the subblock consists of "fused" operations like conv and batch norm?

**Comment by titu1994** [comment]
The paper mentions it as a possibility, say if you trace it and optimize the operations it would be possible to fuse it. However in pytorch at that time, there was no such jit tracer. Nowadays you can do jit trace of pytorch modules and while it will not fuse all the operations, it is feasible that in the future onnx or other optimizing jit compilers can do it automatically v

**Reply by fedorovgv** [reply]
Thanks a lot for the quick and complete answer!
